File: ecs-hosted/go/application/service/insights.go
package service

import (
	// "fmt"

	// "github.com/pulumi/pulumi-aws/sdk/v6/go/aws/ecs"
	"github.com/pulumi/pulumi-self-hosted-installers/ecs-hosted/application/config"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

func DeployOpenSearchDashboards(ctx *pulumi.Context, config *config.ConfigArgs, OpenSearchDomainEndpoint pulumi.StringOutput, vpcId pulumi.StringOutput, subnetIds pulumi.StringArrayOutput, securityGroupId pulumi.StringOutput) error {
	// Create ECS cluster
	// cluster, err := ecs.NewCluster(ctx, "OpenSearchDashboardsCluster", nil)
	// if err != nil {
	// 	return err
	// }

	// // Create ECS task definition
	// taskDefinition, err := ecs.NewTaskDefinition(ctx, "OpenSearchDashboardsTask", &ecs.TaskDefinitionArgs{
	// 	ContainerDefinitions: OpenSearchDomainEndpoint.ApplyT(func(endpoint string) string {
	// 		return fmt.Sprintf(`[
	// 			{
	// 				"name": "OpenSearch-dashboards",
	// 				"image": "OpenSearchproject/OpenSearch-dashboards:latest",
	// 				"memory": %d,
	// 				"cpu": %d,
	// 				"essential": true,
	// 				"portMappings": [
	// 					{
	// 						"containerPort": 5601,
	// 						"hostPort": 5601
	// 					}
	// 				],
	// 				"environment": [
	// 					{
	// 						"name": "OpenSearch_HOSTS",
	// 						"value": "%s"
	// 					}
	// 				]
	// 			}
	// 		]`, config.OpenSearchDashboardsMemory, config.OpenSearchDashboardsCpu, endpoint)
	// 	}).(pulumi.StringOutput),
	// 	Family:      pulumi.String("OpenSearch-dashboards"),
	// 	NetworkMode: pulumi.String("awsvpc"),
	// 	RequiresCompatibilities: pulumi.StringArray{
	// 		pulumi.String("FARGATE"),
	// 	},
	// 	Cpu:              pulumi.String("256"),
	// 	Memory:           pulumi.String("512"),
	// 	ExecutionRoleArn: pulumi.String("arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy"),
	// })
	// if err != nil {
	// 	return err
	// }

	// // Create ECS service
	// _, err = ecs.NewService(ctx, "OpenSearchDashboardsService", &ecs.ServiceArgs{
	// 	Cluster:        cluster.Arn,
	// 	TaskDefinition: taskDefinition.Arn,
	// 	DesiredCount:   pulumi.Int(1),
	// 	NetworkConfiguration: &ecs.ServiceNetworkConfigurationArgs{
	// 		Subnets:        subnetIds,
	// 		SecurityGroups: pulumi.StringArray{securityGroupId},
	// 	},
	// })
	// if err != nil {
	// 	return err
	// }

	return nil
}

File: ecs-hosted/ts/application/services/insights.ts
import * as aws from "@pulumi/aws";
import * as pulumi from "@pulumi/pulumi";
import { config } from "../config";

export class Insights extends pulumi.ComponentResource {
    constructor(name: string, opts?: pulumi.ComponentResourceOptions) {
        super("x:application:insights", name, {}, opts);

        (async () => {
            const callerIdentity = await aws.getCallerIdentity({});
            const openSearchDomain = new aws.opensearch.Domain(`${name}-opensearch`, {
                clusterConfig: {
                    instanceType: config.insights.openSearchInstanceType,
                    instanceCount: config.insights.openSearchInstanceCount,
                },
                ebsOptions: {
                    ebsEnabled: true,
                    volumeSize: config.insights.openSearchVolumeSize,
                },
                accessPolicies: JSON.stringify({
                    Version: "2012-10-17",
                    Statement: [
                        {
                            Effect: "Allow",
                            Principal: "*",
                            Action: "es:*",
                            Resource: pulumi.interpolate`arn:aws:es:${aws.config.region}:${(await callerIdentity).accountId}:domain/${name}-opensearch/*`,
                        },
                    ],
                }),
                vpcOptions: {
                    securityGroupIds: [config.endpointSecurityGroupId],
                    subnetIds: config.privateSubnetIds,
                },
            }, { parent: this });

            const openSearchDashboards = new aws.ecs.Service(`${name}-opensearch-dashboards`, {
                cluster: config.ecsClusterArn,
                taskDefinition: new aws.ecs.TaskDefinition(`${name}-opensearch-dashboards-task`, {
                    family: `${name}-opensearch-dashboards`,
                    containerDefinitions: JSON.stringify([{
                        name: "opensearch-dashboards",
                        image: "opensearchproject/opensearch-dashboards:2.11.1",
                        portMappings: [{ containerPort: 5601 }],
                        environment: [
                            { name: "OPENSEARCH_HOSTS", value: `["http://${openSearchDomain.endpoint}"]` },
                        ],
                    }]),
                    requiresCompatibilities: ["FARGATE"],
                    networkMode: "awsvpc",
                    cpu: "512",
                    memory: "1024",
                }).arn,
                desiredCount: 1,
                launchType: "FARGATE",
                networkConfiguration: {
                    subnets: config.privateSubnetIds,
                    securityGroups: [config.endpointSecurityGroupId],
                },
            }, { parent: this });
        })();
    }
}
